Scala Native

armanbilge opened the pull request:
Closes https://github.com/typelevel/cats-effect/issues/1302.

armanbilge committed:
kernel.native

armanbilge committed:
kernel-testkit.native

armanbilge committed:
laws.native

armanbilge committed:
std.native

armanbilge committed:
core.native compiles

armanbilge committed:
example.native

armanbilge committed:
tests.native compiling

armanbilge committed:
Add missing javalib

armanbilge committed:
Fix FiberMonitor

armanbilge committed:
Fix file name

armanbilge committed:
Use busy-wait scheduler for tests

armanbilge committed:
fix busy wait scheduler

armanbilge committed:
All tests pass

armanbilge committed:
Setup snapshotting

armanbilge committed:
Weird scaladoc issue on native

armanbilge committed:
Fix Scala 3 compile

armanbilge committed:
Merge branch 'typelevel:series/3.x' into experiment/native

armanbilge committed:
Updates

armanbilge committed:
Try setting up CI

armanbilge committed:
scalafmt+scalafix

armanbilge committed:
Add ciNative alias

armanbilge committed:
Explicit types for implicits

armanbilge committed:
De-boot-leg

armanbilge committed:
Fix MiMa

armanbilge committed:
Fix workflow

armanbilge committed:
Implement `SecureRandom` for Native

armanbilge commented on std/native/src/main/scala/cats/effect/std/SecureRandomCompanionPlatform.scala:
```diff
@@ -0,0 +1,56 @@
+
+}
+
+@extern
+@nowarn
+private[std] object sysrandom {
+  def getentropy(buf: Ptr[Byte], buflen: CSize): Int = extern
+}
```
This works on Linux and macOS. I reckon that's plenty good for now. Anyone can implement `SecureRandom` for themselves to suit their requirements.

djspiewak replied:
We definitely need to write up some serious words about the limitations of the current implementation. :-)

armanbilge replied:
Sure. But note that anyone who tries to use this on an unsupported platform will get an error at _linking time_, before runtime. So it's not evil, just annoying.

armanbilge commented on build.sbt:
```diff
@@ -322,12 +345,15 @@ lazy val kernel = crossProject(JSPlatform, JVMPlatform)
   .jsSettings(
     libraryDependencies += "org.scala-js" %%% "scala-js-macrotask-executor" % MacrotaskExecutorVersion % Test
   )
+  .nativeSettings(
+    libraryDependencies += "io.github.cquiroz" %%% "scala-java-time" % "2.4.0"
+  )
```
I don't think Scala Native has the same artifact size concerns as Scala.js, so I reckon we should ship `java.time` support out-of-the-box.

armanbilge committed:
Hopefully docs work now ...

armanbilge committed:
Tweak Native `IOApp` exiting

armanbilge committed:
Use `DetectPlatform` as needed

armanbilge committed:
Scala Native supports `os.version`

armanbilge committed:
Just use global IORuntime for tests

armanbilge commented on core/native/src/main/scala/cats/effect/unsafe/QueueExecutorScheduler.scala:
```diff
@@ -0,0 +1,98 @@
+import java.time.temporal.ChronoField
+
+// inspired by https://github.com/scala-native/scala-native/blob/7058cea339168f9456b8c42f1dc4909653e465e8/nativelib/src/main/scala/scala/scalanative/runtime/ExecutionContext.scala
+private[effect] object QueueExecutorScheduler extends ExecutionContextExecutor with Scheduler {
```
This uses `Thread.sleep()` instead of busy-waiting on the global EC when there's nothing to do. But, it's impractical to use for tests (although _possible_) and it's not especially useful as a real runtime either.

armanbilge replied:
That is to say, I don't really think it's worth keeping.

armanbilge replied:
Thanks to some brainstorming from Daniel this has evolved into a much more interesting `PollingExecutorScheduler` abstraction :)

armanbilge committed:
Fix compile

armanbilge committed:
Extract `PollingExecutorScheduler` abstraction

And implement `QueueExecutorScheduler` in terms of it, and make it the default

armanbilge committed:
Fix Native CI alias

armanbilge committed:
Privatize and finalize stuff

armanbilge committed:
Simplifying

rossabaker commented:
Is there a way to limit the number of concurrent jobs?  Now that the matrix has grown past 20, this is using all of the org's runners.

armanbilge commented:
Yeah, that is indeed annoying 😕

GHA does have some controls for concurrency. I think to make it work though we'd have to split the matrix roughly in half (two batches of 10) and enforce that only one such batch may be running at a time across the entire repo.
https://docs.github.com/en/actions/using-jobs/using-concurrency

On the other hand, this will suck when its crickets everywhere else but we're trying to do something here.

armanbilge commented:
Another alternative I've thought about is running a slimmer CI matrix for PRs and using the full matrix only on merges.

djspiewak commented:
> Is there a way to limit the number of concurrent jobs? Now that the matrix has grown past 20, this is using all of the org's runners.

I'm honestly not sure it can be helped. Cats Effect is often affected by the *other* direction of this problem, btw, where builds on Cats or Fs2 can eat up runners and cause builds here to take a remarkably long time. Gating concurrency doesn't really solve the problem, which is that we can't get more runners without paying a significant premium (and even then we don't get many).

armanbilge committed:
Use `j.u.ArrayDeque` instead of `ListBuffer`

armanbilge committed:
Use `j.u.PriorityQueue`

armanbilge committed:
First pass at better event loop

armanbilge committed:
Fix stupid mistake

armanbilge committed:
Just directly execute 0-sleep tasks

armanbilge committed:
Save some allocations

armanbilge committed:
Fix use of `cachedNow` for `sleep`ing

armanbilge committed:
Less contention on Native

rossabaker commented:
This is a better discussion for the maintainers forum.  This is strangling us all.  I'll move it there.

armanbilge committed:
Merge remote-tracking branch 'upstream/series/3.x' into experiment/native

armanbilge committed:
Re-order steps in event loop

armanbilge committed:
Merge branch 'fix/syncstep-uncancelable-3.x' into experiment/native

armanbilge committed:
Merge branch 'fix/syncstep-uncancelable-3.x' into experiment/native

armanbilge committed:
Formatting

armanbilge committed:
Merge branch 'fix/syncstep-uncancelable-3.x' into experiment/native

armanbilge committed:
Add support for microsecond time

armanbilge committed:
Fix binding

armanbilge committed:
Add fallback for Windows

armanbilge committed:
Make `installGlobal` public

armanbilge commented:
@keynmol kindly reports that all tests are passing on macOS M1 + 2.13 + Native

armanbilge commented on core/native/src/main/scala/cats/effect/unsafe/PollingExecutorScheduler.scala:
```diff
@@ -0,0 +1,181 @@
+        if (sleepIter.next().canceled) sleepIter.remove()
+      }
+
+      // now we poll
+      val timeout =
+        if (!executeQueue.isEmpty())
+          Duration.Zero
+        else if (!sleepQueue.isEmpty())
+          (sleepQueue.peek().at - cachedNow).nanos
+        else
+          Duration.Inf
+
+      val needsPoll = poll(timeout)
+
+      // swap the task queues
+      val todo = executeQueue
+      executeQueue = cachedExecuteQueue
+      cachedExecuteQueue = todo
+
+      // do all the tasks
+      while (!todo.isEmpty()) {
+        val runnable = todo.poll()
+        try {
+          runnable.run()
+        } catch {
+          case NonFatal(t) =>
+            reportFailure(t)
+        }
+      }
+
+      continue = needsPoll || !executeQueue.isEmpty() || !sleepQueue.isEmpty()
```
This is wrong. `needsPoll` may have been false, but then we run a bunch of tasks that may schedule new stuff for `poll`ing. Indeed, we may have drained the queue. So we could end up not continuing, when in fact we should be `poll`ing.

armanbilge committed:
Add explanation doc b/c, well, even I forgot (:

armanbilge committed:
poll as the last step in the loop

armanbilge committed:
Cleanup canceled timers before polling

armanbilge committed:
Use secure UUIDGen for Native

armanbilge committed:
Bump Scala Native version

armanbilge committed:
Surround blocking calls with cede

armanbilge commented on core/native/src/main/scala/cats/effect/IOCompanionPlatform.scala:
```diff
@@ -23,11 +23,12 @@ import java.time.Instant
   def blocking[A](thunk: => A): IO[A] =
-    apply(thunk)
+    // do our best to mitigate blocking
+    IO.cede *> apply(thunk) <* IO.cede
```
Downstream libraries are riddled with blocking calls (e.g. DNS resolution in ip4s, file I/O in fs2). Actually even `IO.println` is blocking! Since we don't have a blocking threadpool to run them on, this seems like the best we can do to mitigate their effects.

armanbilge replied:
Actually we can do even better here: this should be `apply(thunk).guarantee(IO.cede)`. We don't want to skip that `cede` even if we errored.

armanbilge committed:
Update `nowMicros` with SN 0.4.7 APIs

armanbilge committed:
More cleanup for `nowMicros`

armanbilge committed:
`guarantee(cede)` after `blocking` `thunk`

armanbilge committed:
Add failing test for `SecureRandom#nextInt`

armanbilge committed:
Correctly override `SecureRandom#next`

armanbilge committed:
Gracefully handle negative sleeps

armanbilge committed:
Merge remote-tracking branch 'upstream/series/3.x' into experiment/native

armanbilge committed:
Add `pollEvery` parameter for fairness, default 64

armanbilge committed:
`try`/`catch` like the WSTP

armanbilge committed:
`ScheduledTask` -> `SleepTask`

armanbilge committed:
Refactor `sleep` cancelation

Now behaves like `ScheduledExecutorService#setRemoveOnCancelPolicy(true)` used on JVM

armanbilge committed:
Rip out `cachedNow` stuff

armanbilge committed:
Revert "Make `installGlobal` public"

This reverts commit 7efbac0eb49239252b556db5c0b7abaa11997ea5.

armanbilge committed:
Fix `suspend` impl

djspiewak commented:
So this is all in `series/3.3.x` now right? How do we untangle the git gordian knot?

armanbilge commented:
No, this is not really in `series/3.3.x`. Actually pulling that off would have been too painful because Scala Native requires Scala 3.1+ and the JVM/JS `series/3.3.x` is stuck on Scala 3.0. Hence my "backport" PR https://github.com/typelevel/cats-effect/pull/3138 really started a parallel branch `series/3.3.x-native`. That branch can simply be abandoned.

Yes, sorry, not ideal. But I wasn't prepared to let this die-on-the-vine like my `Env` PR from \*checks watch\* ...

djspiewak commented:
> Yes, sorry, not ideal. But I wasn't prepared to let this die-on-the-vine like my Env PR from *checks watch* ...

*whistles innocently*

Either way, given that this has now been released, I think it's best if we get it into series/3.x to form a part of 3.4. :-)

armanbilge committed:
Merge remote-tracking branch 'upstream/series/3.x' into experiment/native

armanbilge commented:
Ok, @djspiewak, let's land this one :) tbh I was quite sad to ship this out in 3.3.14 without your review, so I'm looking forward to that 😇 

Question: do we want to add a macOS CI job? There a couple places where we directly call native APIs (high-precision time and cryptographically strong random numbers). They are all supported on Linux/macOS, but in theory we should be careful.

djspiewak commented:
> Question: do we want to add a macOS CI job? There a couple places where we directly call native APIs (high-precision time and cryptographically strong random numbers). They are all supported on Linux/macOS, but in theory we should be careful.

I hate to say it, but we should be careful. :-( So that means one more matrix entry.

armanbilge committed:
Add native+macOS job

armanbilge commented:
Something I haven't done in this PR is setup the test harness so we can run the `IOAppSpec` with Native executables. It's possible by the same trick we use to run the `IOAppSpec` for JS. So just a yak I'd be happy to put off 😆

djspiewak commented on core/native/src/main/scala/cats/effect/IOApp.scala:
```diff
@@ -0,0 +1,244 @@
+    // it isn't silently discarded, as longer ticks are, but slow
+    // enough that we don't interrupt often.  1 hour was chosen
+    // empirically.
+    lazy val keepAlive: IO[Nothing] =
+      IO.sleep(1.hour) >> keepAlive
```
Do we actually need this with Scala Native?

armanbilge replied:
I believe so! The whole point of this is to keep an `IO.never` program alive. In the `IO.never` case the EC queue will be empty. In theory we could just have the main thread go into a permanently parked state in this situation.

But in practice the EC actually has to stop running if its queues are empty so that it can yield to the Scala Native global EC, to make test frameworks work correctly. See also the doc comment on `poll(...)`.

djspiewak replied:
Fascinating. It's identical code but actually handled in an entirely different way.

djspiewak commented on core/native/src/main/scala/cats/effect/IOCompanionPlatform.scala:
```diff
@@ -0,0 +1,65 @@
+
+  private[this] val TypeDelay = Sync.Type.Delay
+
+  def blocking[A](thunk: => A): IO[A] =
+    // do our best to mitigate blocking
+    IO.cede *> apply(thunk).guarantee(IO.cede)
```
😱

djspiewak commented on core/native/src/main/scala/cats/effect/IOApp.scala:
```diff
@@ -0,0 +1,244 @@
+import scala.concurrent.CancellationException
+import scala.concurrent.duration._
+
+/**
```
Can we write some words in here which clarifies 1) the experimental nature of things, and 2) the limitations imposed by this particular `IOApp`? It's fair to call out to EpollCat for now since that's the only usable runtime.

armanbilge replied:
Yep can write some doc. Besides the `blocking` issue I wouldn't really say `IOApp` is limited. Anything you can do in Scala Native today works fine with this `IOApp`. If you want to do async I/O with JDK APIs, tough, because they are not implemented anyway, but that's not this `IOApp`s fault.

> It's fair to call out to EpollCat for now since that's the only usable runtime.

It's not actually. It's only necessary for `fs2.io.net` and thus ember or skunk etc.

I have a `CurlApp` in [http4s-curl](https://github.com/http4s/http4s-curl) that's perfectly good if the only non-blocking I/O you need are HTTP requests (this is sufficient to implement a Lambda runtime btw). Also that runtime is theoretically capable of polling on arbitrary file descriptors, but I have not exposed that.

I also just wrote an `io_uring`-based runtime in https://github.com/armanbilge/fs2-io_uring/pull/2. But that's still a bit premature :)

Finally, see also https://github.com/typelevel/cats-effect/pull/3153#discussion_r974603033 where we discussed some interesting native applications built with vanilla `IOApp`, including CLIs and an LSP.

djspiewak commented on core/native/src/main/scala/cats/effect/tracing/TracingPlatform.scala:
```diff
@@ -0,0 +1,34 @@
+
+private[tracing] abstract class TracingPlatform { self: Tracing.type =>
+
+  @nowarn("msg=never used")
+  def calculateTracingEvent(key: Any): TracingEvent = null
```
This will be fun™…

djspiewak commented on tests/native/src/test/scala/cats/effect/DetectPlatform.scala:
```diff
@@ -0,0 +1,23 @@
+package cats.effect
+
+trait DetectPlatform {
+  def isWSL: Boolean = System.getProperty("os.version").contains("-WSL")
```
How are system properties handled on Scala Native?

armanbilge replied:
No idea what this would say for WSL. But I believe it does attempt to populate them as the JVM would.

djspiewak reviewed (commented):
Mostly just worried about the lack of documentation and caveats!

armanbilge committed:
Add implementation note/limits for `SecureRandom`

armanbilge committed:
Document limitations for `IOApp` on Native
